src/state_machine.zig

```diff
@@ -393,35 +393,43 @@ pub fn StateMachineType(
             cache_entries_account_balances: u32,
         };
 
-        /// Since prefetch contexts are used one at a time, it's safe to access
-        /// the union's fields and reuse the same memory for all context instances.
-        const PrefetchContext = union(enum) {
-            null,
-            accounts: AccountsGroove.PrefetchContext,
-            transfers: TransfersGroove.PrefetchContext,
-            transfers_pending: TransfersPendingGroove.PrefetchContext,
-
-            pub const Field = std.meta.FieldEnum(PrefetchContext);
-            pub fn FieldType(comptime field: Field) type {
-                return std.meta.fieldInfo(PrefetchContext, field).type;
+        const PrefetchContext = struct {
+            const Fields = struct {
+                accounts: AccountsGroove.PrefetchContext,
+                transfers: TransfersGroove.PrefetchContext,
+                transfers_pending: TransfersPendingGroove.PrefetchContext,
+            };
+            const Tag = std.meta.FieldEnum(Fields);
+            const EnumSet = std.EnumSet(Tag);
```

Comment:
If we are introducing a name here, could we also up the abstraction layer here?

```
used: FieldSet
```

Reply:
Hm, wait, I think there's a bigger picture here!

The smell right now is that the new PrefetchContext encapsulates "type dispatch" logic, where we do something meta to get at accounts/transfers/pending. But we already have this logic encoded in the Forest!

So can we sink `PrefetchContext` to be a field of `Groove` (I think we can statically `void` the fild if `prefetch_entries_for_read_max == 0`, like it is for accounts_balances groove)?

I think this _wasn't_ done before exactly to take advantage of this union memory sharding, but we can't do this anymore (and it seems like there isn't actually that much memory to share here?).

However, there's one thing which we _must_ share --- 

```
workers: [Grid.read_iops_max]PrefetchWorker = undefined,
```

These read_iops must be shared across all concurrent prefetches. 

Well, maybe not literally the workers, because workers are type-specific, but the underlying Grid.Reads definietly should be shared.

So it seems like we should keep reads at the state machine level (or maybe Forest levl, for better encapsulation), but sink the rest of prefetch context down to individual grooves?

Reply:
(to clarify, the sinking types part here is feng-shui, and maybe we don't actually need that. But sharing the pool of reads across all prefetches is observrable behavior, and something I think we should tackle)

Reply:
> Hm, wait, I think there's a bigger picture here!

Yes, I have plans to unify prefetch lookups and [ScanLookupType](https://github.com/tigerbeetle/tigerbeetle/blob/bc4a71bc5fa93de1f33403b0ea1d7a1f176e565e/src/lsm/scan_lookup.zig#L22), and the idea of a per-forest pool of resources is pretty much what I had in mind.